lib/propmangle.js

```diff
@@ -159,7 +162,7 @@ function mangle_properties(ast, options) {
         if (options.only_cache) {
             return cache.props.has(name);
         }
-        if (/^[0-9.]+$/.test(name)) return false;
+        if (/^-?[0-9]+(\.[0-9]+)?(e[+-][0-9]+)?$/.test(name)) return false;
```

Comment:
PTAL at https://github.com/mishoo/UglifyJS2/pull/1779/commits/b78f0554a3dcb79d0402146b6ea7294ad2eae0a9 and see if the new test aligns with what you expect to see.

Specifically, if you want `[Ee][+-]?` then the `"1E42"` case will need to be updated.